cpp/include/cudf/wrappers/durations.hpp

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <limits>
+
+#define _LIBCUDACXX_USE_CXX20_CHRONO
+#define _LIBCUDACXX_USE_CXX17_TYPE_TRAITS
+
+#include <simt/chrono>
+
+/**
+ * @file durations.hpp
+ * @brief Concrete type definitions for int32_t and int64_t durations in varying resolutions.
+ **/
+namespace cudf {
+namespace detail {
+
+template <class Duration>
+struct duration : Duration {
```

Comment:
The duration helper types like `std::chrono::days` don't provide sufficient guarantees for the size of the type, e.g., `std::chrono::days` is only guaranteed to be at least 25 bits. 

That said, I can't recall why we made the explicit `timestamp` type instead of doing:
```
using timestamp_D = simt::std::chrono::time_point<simt::std::chrono::system_clock, simt::std::chrono::duration<int32_t, simt::std::ratio<86400>>>
```

@trxcllnt do you remember why we didn't make the libcudf timestamp types simply be aliases and instead made a concrete new type? I'm sure we had a good reason, but I can't remember right now.

Reply:
Yeah I think there were a few things:

1. We didn't want to make the clock part of the type, since we expected to switch to `utc_clock` eventually once it lands in libcu++.
2. We have an extra constructor that takes a numeric `typename Duration::rep r` type that's more convenient to use
3. For some reason we needed to override the conversion constructor to explicitly call `duration_cast()` since the [libcu++ impl](https://github.com/rapidsai/thirdparty-libcxx/blob/a97a7380c76346c22bb67b93695bed19592afad2/include/chrono#L1383) doesn't call it (possible bug in libcu++?).

Reply:
for the 2nd item above - should we be even supporting constructing a timestamp with a tick count? in this case, what does 1000 really represent? shouldn't it be a duration (of ms in this example) or another time_point with a similar or a different duration. presently, we interpret to mean add 1000 ms to the system_clock epoch (as we only support system_clock for now), which doesn't look to be the original intent, as 1000 could mean different units or displacement from some other clock.

i think 3rd item is required if we want to truncate and lose precision - i.e. assigning say `timestamp_s` to a `timestamp_D`. that constructor will prevent this if the periods aren't divisible - `is_convertible` should evaluate to true.

Reply:
2). 2nd item is present for ease the testing code. `timestamp(typename Duration::rep r)` should be **explicit**. 
3) 3rd item is not bug since `is_convertible` takes care of it. (lossy conversion is not allowed as per `std::chrono`, needs explicit `duration_cast`). `std::chrono` does not recommend using `duration_cast` (lossy conversion) implicitly. One plus point is that it's explicit constructor.
Perhaps lossy conversion is the requirement from users (Spark, Python, etc). is it?

Reply:
@karthikeyann do you think we should move away from having our own concrete `timestamp` type and instead just us a type alias for both timestamps and durations?

Reply:
We can't replace due to a CUDA limitation. If `timestamp` is replaced with `simt::std::chrono::time_point`, following error is hit due to limitation of shared memory initialization. 
`cpp/include/cudf/detail/copy_if.cuh(106): error: initializer not allowed for __shared__ variable `
https://github.com/rapidsai/cudf/blob/efebcd1452692ee15f3d30627a9ef3d0cafa85d5/cpp/include/cudf/detail/copy_if.cuh#L106

This is a limitation of device variable. https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#device-memory-specifiers
- The constructor function has no parameters, the initializer list is empty and the function body is an empty compound statement.
- The default constructors of all base classes of its class can be considered empty.

`simt::std::chrono::time_point` calls parameterized constructor of _duration_
https://github.com/rapidsai/thirdparty-libcxx/blob/a97a7380c76346c22bb67b93695bed19592afad2/include/chrono#L1372

`    _LIBCPP_INLINE_VISIBILITY _LIBCPP_CONSTEXPR_AFTER_CXX11 time_point() : __d_(duration::zero()) {}`

`timestamp` class avoids it by calling no parameter constructor of _duration_.

Reply:
`simt::std::chrono::duration` is used for duration types. 
for timestamp, wait for #5400 .